### PR TITLE
Update. DASH-107 Ref/Contained flow updates

### PR DESCRIPTION
**Pull request by trukhilio**
- Use regular SignNFT page for SignNFT contained flow
- Delete ContainedFlowActionPage
- Handle fioActionExecuted results
- Create ContainedFlowWrapper for MainLayout
- Enable get ref params from query
- Add new REG contained flow
- Refactor contained flow using steps
- Delete containedFlow hook
- Fix redirect on email block page and confirm email double call issue
- Create Account UI fixes, Badge UI fix
- Fix redirect on checkout page if multiple wallets
- Hide AddressDomainBadge on contained flow
- Fix titles/subtitles values for contained flow

### CODE REVIEW

File: client/src/components/AddressDomainContainer/AddressDomainContainer.tsx [hunk 1]
```diff
@@ -50,7 +52,9 @@ const AddressDomainContainer: React.FC<Props &
             hasFreeAddress={hasFreeAddress}
           />
           {!isDesktop && <AddressDomainCart />}
-          {allowCustomDomains && <AddressDomainBadge type={type} />}
+          {allowCustomDomains && !isContainedFlow && (
```
Comment (andreyvEze) [inline client/src/components/AddressDomainContainer/AddressDomainContainer.tsx:55]: should it be setup from ref profile? why we hide it when contained flow?

File: client/src/components/AddressDomainContainer/AddressDomainContainer.tsx [hunk 2]
```diff
@@ -50,7 +52,9 @@ const AddressDomainContainer: React.FC<Props &
...
+            <AddressDomainBadge type={type} />
+          )}
```
Comment (andreyvEze) [inline client/src/components/AddressDomainContainer/AddressDomainContainer.tsx:57]: also, should it affect somehow the widget?

File: client/src/components/CheckoutPurchaseContainer/CheckoutPurchaseContainer.tsx [hunk 3]
```diff
@@ -65,6 +68,20 @@ const CheckoutPurchaseContainer: React.FC<Props> = props => {
...
     setProcessing(false);
+    if (!isRetry) {
+      const txIdsArr: string[] = [];
+      results.registered.forEach(regAddress => {
+        const { transaction_id, domainTxId } = regAddress;
+        if (transaction_id) txIdsArr.push(transaction_id);
+        if (domainTxId) txIdsArr.push(domainTxId);
+      });
+
+      fioActionExecuted({
```
Comment (andreyvEze) [inline client/src/components/CheckoutPurchaseContainer/CheckoutPurchaseContainer.tsx:79]: let's check if this should be called not only when !isRetry

File: client/src/components/EdgeConfirmAction/EdgeConfirmAction.tsx [hunk 4]
```diff
@@ -72,8 +72,10 @@ const EdgeConfirmAction: React.FC<Props> = props => {
 
           onSuccess(result);
           fioActionExecuted({
-            ...result,
+            result,
+            additionalData,
             executeActionType: removeExtraCharactersFromString(action),
+            txIds: result?.transaction_id,
```
Comment (andreyvEze) [inline client/src/components/EdgeConfirmAction/EdgeConfirmAction.tsx:78]: proposition: let's check `action` and call `fioActionExecuted` only when it was fio action

File: client/src/components/EdgeConfirmAction/EdgeConfirmAction.tsx [hunk 5]
```diff
@@ -72,8 +72,10 @@ const EdgeConfirmAction: React.FC<Props> = props => {
 
           onSuccess(result);
           fioActionExecuted({
-            ...result,
+            result,
```
Comment (andreyvEze) [inline client/src/components/EdgeConfirmAction/EdgeConfirmAction.tsx:75]: let's discuss result types and this method params type

File: client/src/components/PurchaseNow/middleware.ts [hunk 6]
```diff
@@ -274,6 +274,8 @@ const handleResponses = (
             result.registered[existingCartItemRegisteredIndex],
             response.value,
           ),
+          domainTxId:
```
Comment (andreyvEze) [inline client/src/components/PurchaseNow/middleware.ts:277]: why we need to separate it?

File: client/src/components/SignNft/SignNft.tsx [hunk 7]
```diff
@@ -86,6 +87,14 @@ const SignNft: React.FC<ContainerProps> = props => {
     }
   }, []);
 
+  const initRes = JSON.stringify(initialResults);
```
Comment (andreyvEze) [inline client/src/components/SignNft/SignNft.tsx:90]: need to discuss

File: client/src/components/SignNft/SignNft.tsx [hunk 8]
```diff
@@ -116,21 +126,9 @@ const SignNft: React.FC<ContainerProps> = props => {
...
     if (result != null) {
-      const { metadata, ...rest } = result.other.nfts[0];
-      const creatorUrl = (() => {
-        try {
-          return JSON.parse(metadata || '').creator_url;
-        } catch (err) {
-          return '';
-        }
-      })();
-
```
Comment (andreyvEze) [inline client/src/components/SignNft/SignNft.tsx:127]: want to clarify why do we need this change

File: client/src/redux/containedFlow/sagas.ts [hunk 9]
```diff
@@ -35,22 +40,43 @@ export function* fioAddressRegisterSuccess(): Generator {
...
+            routeParams = ROUTES.PURCHASE;
+            break;
+          default:
+            routeParams = ROUTES.HOME;
+        }
+
+        history.push(
+          routeParams,
+          action.data.result, // pass results to state
+        );
```
Comment (andreyvEze) [inline client/src/redux/containedFlow/sagas.ts:70]: let's discuss why we need to navigate here

File: client/src/redux/types.d.ts [hunk 10]
```diff
@@ -50,4 +61,5 @@ export type Action = {
   PricesAction &
   CommonPromiseAction &
   CommonAction &
-  any;
+  ContainedFlowAction &
+  AnyType; // todo: check why do we need this here
```
Comment (andreyvEze) [inline client/src/redux/types.d.ts:65]: we need to set proper types somtimes - all possible types I assume

File: client/src/services/ContainedFlow.tsx [hunk 11]
```diff
@@ -65,4 +65,4 @@ const reduxConnect = connect(
   },
 );
 
-export default compose(reduxConnect)(ContainedFlow);
+export default withRouter(compose(reduxConnect)(ContainedFlow));
```
Comment (andreyvEze) [inline client/src/services/ContainedFlow.tsx:68]: what is wrong with `useHistory`?

File: client/src/util/homePage.ts [hunk 12]
```diff
@@ -41,10 +41,12 @@ export const handleHomePageContent = ({
...
         actions[actionName as ContainedFlowActionSettingsKey]?.title ||
-        CONTAINED_FLOW_TITLES[actionName];
+        CONTAINED_FLOW_TITLES[actionName] ||
+        '';
       subtitle =
-        actions[actionName as ContainedFlowActionSettingsKey].subtitle ||
-        CONTAINED_FLOW_SUBTITLES[actionName];
+        actions[actionName as ContainedFlowActionSettingsKey]?.subtitle ||
+        CONTAINED_FLOW_SUBTITLES[actionName] ||
+        '';
```
Comment (andreyvEze) [inline client/src/util/homePage.ts:49]: if empty string it will set some other default title in the end?

**Review by andreyvEze — CHANGES_REQUESTED** [review]
review

File: client/src/components/AddressDomainContainer/AddressDomainContainer.tsx (hunk 2 above)
Comment (andreyvEze) [inline client/src/components/AddressDomainContainer/AddressDomainContainer.tsx:57]: could you please change `AddressDomainFrom` -> `AddressDomainForm`

File: client/src/components/EdgeConfirmAction/EdgeConfirmAction.tsx (hunk 4 above)
Comment (andreyvEze) [inline client/src/components/EdgeConfirmAction/EdgeConfirmAction.tsx:78]:
```
{
 result,
 data,
 executeActionType
}
```

File: client/src/components/SignNft/SignNft.tsx (hunk 7 above)
Comment (andreyvEze) [reply client/src/components/SignNft/SignNft.tsx:90]: assume it would be removed

File: client/src/redux/containedFlow/sagas.ts (hunk 9 above)
Comment (andreyvEze) [reply client/src/redux/containedFlow/sagas.ts:70]: will be removed and created new saga for Results Close event

File: client/src/services/ContainedFlow.tsx (hunk 11 above)
Comment (andreyvEze) [reply client/src/services/ContainedFlow.tsx:68]: let's turn it back for now, please update todo message

File: client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx [hunk 13]
```diff
@@ -1,29 +1,48 @@
...
 import { putParamsToUrl } from '../../utils';
 import { ROUTES } from '../../constants/routes';
 
-type MatchParams = {
-  address: string;
-};
+import { ContainedFlowQueryParams } from '../../types';
+
+const FioAddressSignPage: React.FC = () => {
+  const params: { address?: string } = useParams();
```
Comment (andreyvEze) [inline client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx:11]: why we need the hook if we could get params from props

File: client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx [hunk 14]
```diff
@@ -1,29 +1,48 @@
...
   return (
     <>
       <SignNft
-        addressSelectOff={true}
+        initialValues={initialValues}
+        addressSelectOff={address}
         fioAddressName={address}
-        backTo={putParamsToUrl(ROUTES.FIO_ADDRESS_SIGNATURES, { address })}
+        backTo={
+          containedFlowQueryParams
```
Comment (andreyvEze) [inline client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx:42]: here we should check for `address`

Comment (andreyvEze) [inline client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx:42]: here we should check for `address`

File: client/src/constants/common.ts [hunk 15]
```diff
@@ -12,30 +12,35 @@ export const DEFAULT_WALLET_OPTIONS = {
   fiatCurrencyCode: 'iso:USD',
 };
 export const FIO_WALLET_TYPE = 'wallet:fio';
-export const CONFIRM_PIN_ACTIONS = {
-  RECOVERY: 'RECOVERY',
+
+export const FIO_ACTIONS = {
```
Comment (andreyvEze) [inline client/src/constants/common.ts:16]: `CONFIRM_PIN_FIO_ACTIONS`

File: client/src/components/common/TransactionResults/Results.tsx [hunk 16]
```diff
@@ -34,13 +34,19 @@ const Results: React.FC<ResultsContainerProps> = props => {
...
+    setContainedFlowFinishStep,
     children,
   } = props;
 
   useEffect(() => {
     window.scrollTo(0, 0);
   }, []);
 
+  const handleClose = () => {
+    setContainedFlowFinishStep();
```
Comment (andreyvEze) [inline client/src/components/common/TransactionResults/Results.tsx:46]: do not use contained flow logic, here should be some event called like `txResultsClosed`

File: client/src/pages/PurchasePage/PurchasePage.tsx [hunk 17]
```diff
@@ -76,9 +82,13 @@ const PurchasePage: React.FC<Props & RouteComponentProps> = props => {
...
-      return history.push(
-        CONTAINED_FLOW_ACTIONS_TO_ROUTES[containedFlowQueryParams.action],
-      );
+      if (containedFlowQueryParams.action === CONTAINED_FLOW_ACTIONS.REG) {
+        setStep(CONTAINED_FLOW_STEPS.FINISH);
+      } else {
+        return setStep(CONTAINED_FLOW_STEPS.ACTION, {
+          containedFlowAction: containedFlowQueryParams.action,
+        });
+      }
```
Comment (andreyvEze) [inline client/src/pages/PurchasePage/PurchasePage.tsx:91]: maybe would be cool if we do here the same as on resultsClose

File: client/src/redux/containedFlow/actions.ts [hunk 18]
```diff
@@ -69,7 +72,26 @@ export const setContainedParams = (query: ContainedFlowQuery): CommonAction => {
...
-  data: step,
+  data,
+  step,
+});
+
+export const FINISH_CONTAINED_FLOW = `${prefix}/FINISH_CONTAINED_FLOW`;
+
+export const setContainedFlowFinishStep = (): CommonAction => ({
+  type: FINISH_CONTAINED_FLOW,
+});
```
Comment (andreyvEze) [inline client/src/redux/containedFlow/actions.ts:91]: should be removed/moved to other actions

File: client/src/redux/containedFlow/sagas.ts [hunk 19]
```diff
@@ -1,56 +1,102 @@
...
+
+        yield put(setStep(CONTAINED_FLOW_STEPS.FINISH, { redirectUrl }));
+      }
     }
   });
 }
 
-export function* fioAddressRegisterSuccess(): Generator {
-  yield takeEvery(SET_REGISTRATION_RESULTS, function*(action: Action) {
+export function* finishContainedFlow(): Generator {
```
Comment (andreyvEze) [inline client/src/redux/containedFlow/sagas.ts:55]: `resultsCloseContainedFlow`

File: client/src/redux/containedFlow/sagas.ts [hunk 20]
```diff
@@ -1,56 +1,102 @@
...
+  yield takeEvery(FINISH_CONTAINED_FLOW, function*() {
```
Comment (andreyvEze) [inline client/src/redux/containedFlow/sagas.ts:56]: should be other name

File: client/src/services/ContainedFlow.tsx [hunk 21]
```diff
@@ -5,52 +5,89 @@ import { createStructuredSelector } from 'reselect';
...
   }, [containedFlowQueryParams, query, setContainedParams]);
 
   useEffect(() => {
-    if (actionCondition)
-      history.push(CONTAINED_FLOW_ACTIONS_TO_ROUTES[containedFlowAction]);
-  }, [actionCondition, history, containedFlowAction]);
+    if (isContainedFlow && isAuthenticated) {
+      if (containedFlowAction !== CONTAINED_FLOW_ACTIONS.REG) {
+        if (fioAddresses.length) {
+          return setStep(CONTAINED_FLOW_STEPS.ACTION, { containedFlowAction });
```
Comment (andreyvEze) [inline client/src/services/ContainedFlow.tsx:76]:
```
return setStep(CONTAINED_FLOW_STEPS.ACTION, { initialValues: containedFlowAction });
```

File: client/src/pages/CheckoutPage/CheckoutPage.tsx [hunk 22]
```diff
@@ -86,10 +86,11 @@ const CheckoutPage: React.FC<Props> = props => {
...
       !isFree &&
-      paymentWalletPublicKey &&
-      new MathOp(walletBalancesAvailable.nativeFio).lt(
-        totalCost(cartItems, roe).costNativeFio,
-      )
+      ((paymentWalletPublicKey &&
+        new MathOp(walletBalancesAvailable.nativeFio).lt(
+          totalCost(cartItems, roe).costNativeFio,
+        )) ||
+        (!paymentWalletPublicKey && fioWallets.length > 1))
```
Comment (andreyvEze) [inline client/src/pages/CheckoutPage/CheckoutPage.tsx:93]: discuss

**Review by andreyvEze — COMMENTED** [review]
check comments

File: client/src/constants/common.ts (hunk 15 above)
Comment (trukhilio) [reply client/src/constants/common.ts:16]: done

File: client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx (hunk 14 above)
Comment (trukhilio) [reply client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx:42]: done

File: client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx [hunk 23]
```diff
@@ -1,29 +1,48 @@
...
->> = props => {
   const {
-    match: {
-      params: { address },
-    },
-  } = props;
+    state,
+  }: {
+    state: {
+      containedFlowQueryParams?: ContainedFlowQueryParams;
```
[the remark this answers is not in the document]
Comment (trukhilio) [reply client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx:18]: done

File: client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx (hunk 13 above)
Comment (trukhilio) [reply client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx:11]: done

File: client/src/redux/containedFlow/sagas.ts (hunk 19 above)
Comment (trukhilio) [reply client/src/redux/containedFlow/sagas.ts:55]: done

File: client/src/redux/containedFlow/sagas.ts (hunk 20 above)
Comment (trukhilio) [reply client/src/redux/containedFlow/sagas.ts:56]: done

File: client/src/services/ContainedFlow.tsx (hunk 21 above)
Comment (trukhilio) [reply client/src/services/ContainedFlow.tsx:76]: it should be as is, initial values are on the other place

File: client/src/pages/EmailConfirmationResultsPage/EmailConfirmationResultsPage.tsx [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+    emailConfirmationResult,
+    cartItems,
+    showLoginModal,
+    logout,
+    resetLastAuthData,
+    resetEmailConfirmationResult,
+    createNotification,
+  } = props;
+
+  const history = useHistory();
```
[the remark this answers is not in the document]
Comment (trukhilio) [reply client/src/pages/EmailConfirmationResultsPage/EmailConfirmationResultsPage.tsx:52]: done

File: client/src/components/common/TransactionResults/Results.tsx (hunk 16 above)
Comment (trukhilio) [reply client/src/components/common/TransactionResults/Results.tsx:46]: done

File: client/src/pages/PurchasePage/PurchasePage.tsx (hunk 17 above)
Comment (trukhilio) [reply client/src/pages/PurchasePage/PurchasePage.tsx]: done

File: client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx [hunk 25]
```diff
@@ -1,34 +1,38 @@
...
-    state,
-  }: {
-    state: {
-      containedFlowQueryParams?: ContainedFlowQueryParams;
-    };
-  } = location;
+    match: {
+      params: { address },
+    },
+    location: { state },
```
Comment (andreyvEze) [inline client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx:21]: is it there? just to confirm

File: client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx [hunk 26]
```diff
@@ -39,7 +43,7 @@ const FioAddressSignPage: React.FC = () => {
         addressSelectOff={address}
         fioAddressName={address}
         backTo={
-          containedFlowQueryParams
+          !address
```
Comment (andreyvEze) [inline client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx:46]: could we change the order:
```
address ? putParamsToUrl(ROUTES.FIO_ADDRESS_SIGNATURES, { address }) : null
```

File: client/src/pages/PurchasePage/PurchasePage.tsx [hunk 27]
```diff
@@ -81,17 +74,7 @@ const PurchasePage: React.FC<Props & RouteComponentProps> = props => {
...
-        setStep(CONTAINED_FLOW_STEPS.FINISH);
-      } else {
-        return setStep(CONTAINED_FLOW_STEPS.ACTION, {
-          containedFlowAction: containedFlowQueryParams.action,
-        });
-      }
-    }
-
-    history.push(ROUTES.HOME);
+    onPurchaseResultsClose();
```
Comment (andreyvEze) [inline client/src/pages/PurchasePage/PurchasePage.tsx:77]: 👍

File: client/src/redux/registrations/sagas.ts [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+
+import {
+  CONTAINED_FLOW_STEPS,
+  CONTAINED_FLOW_ACTIONS,
+} from '../../constants/containedFlow';
+import { ROUTES } from '../../constants/routes';
+
+import { ContainedFlowQueryParams } from '../../types';
+
+export function* purchaseResultsClose(history: History): Generator {
```
Comment (andreyvEze) [inline client/src/redux/registrations/sagas.ts:20]: move this saga to contained flow sagas

**Review by andreyvEze — CHANGES_REQUESTED** [review]
please change last comments

File: client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx (hunk 26 above)
Comment (trukhilio) [reply client/src/pages/FioAddressSignPage/FioAddressSignPage.tsx:46]: Done

**Pull request merged**